source/v5/mqtt5_to_mqtt3_adapter.c

```diff
@@ -2226,23 +2234,22 @@ static void s_aws_mqtt5_to_mqtt3_adapter_subscribe_completion_fn(
             "id=%p: mqtt3-to-5-adapter, completing multi-topic subscribe",
             (void *)adapter);
 
+        AWS_VARIABLE_LENGTH_ARRAY(struct aws_mqtt_topic_subscription, multi_sub_subscription_buf, reason_code_count);
         AWS_VARIABLE_LENGTH_ARRAY(
-            struct aws_mqtt_topic_subscription, multi_sub_subscription_buf, suback->reason_code_count);
-        AWS_VARIABLE_LENGTH_ARRAY(
-            struct aws_mqtt_topic_subscription *, multi_sub_subscription_ptr_buf, suback->reason_code_count);
+            struct aws_mqtt_topic_subscription *, multi_sub_subscription_ptr_buf, reason_code_count);
         struct aws_mqtt_topic_subscription *subscription_ptr =
             (struct aws_mqtt_topic_subscription *)multi_sub_subscription_buf;
 
         struct aws_array_list multi_sub_list;
         aws_array_list_init_static(
             &multi_sub_list,
             multi_sub_subscription_ptr_buf,
-            suback->reason_code_count,
+            reason_code_count,
```

Comment:
`aws_array_list_init_static` has a check for the third parameter:
```c
    AWS_FATAL_PRECONDITION(item_count > 0);
```

It seems we should handle `reason_code_count == 0` separately.

Reply:
It sounds like a null suback may be normal and expected behavior. I don't think we want to fail the `AWS_FATAL_PRECONDITION` on something we're doing intentionally (e.g. op timeout, close connection). We may want to set the `subscription_count` earlier and use it since it's grabbing that from the stored `&subscribe_op->subscriptions)` array length on L2250.

Reply:
it is an expected behavior, this function is called with an explicit null as the suback

Reply:
Looking at this case, we would need a test for suback on multiple subscription as well.

Reply:
Ah right, maybe we need a test that sets the multi suback but not the singular one so it hits the multi processing.